File: PROGRAMMERS/LEVEL_02/Q49993.cpp
//https://programmers.co.kr/learn/courses/30/lessons/49993
//스킬트리
#include <iostream>

#include <string>
#include <vector>

using namespace std;

// 각 문자열의 스킬 순서가 가능한 스킬트리인지 검사
bool possible(int input_order[], string &tree)
{
    int t_order = 1;
    for (int i = 0; i < tree.size(); ++i)
    {
        // 선행스킬 목록에 주어진 스킬만 검사
        if (input_order[tree[i] - 'A'])
        {
            // 선행스킬과 순서가 다르면 false 리턴
            if (input_order[tree[i] - 'A'] != t_order)
            {
                return false;
            }
            // 같으면 인덱스 증가
            ++t_order;
        }
    }

    return true;
}

int solution(string skill, vector<string> skill_trees) {
    
    // 선행스킬로 주어진 스킬트리 순서 입력
    const int MAX = 26;
    int input_order[MAX] = { 0, };
    int order = 1;
    for (int i = 0; i < skill.size(); ++i)
    {
        input_order[skill[i] - 'A'] = order++;
    }
    
    // 각 스킬 트리 검사
    int answer = 0;
    for (int i = 0; i < skill_trees.size(); ++i)
    {
        if (possible(input_order, skill_trees[i]))
        {
            ++answer;
        }
    }

    return answer;
}

int main()
{
    string skill = "CBD";
    vector<string> skill_trees = { "BACDE", "CBADF", "AECB", "BDA" };

    cout << solution(skill, skill_trees);

    return 0;
}

File: PROGRAMMERS/LEVEL_02/Q49993.java
package PROGRAMMERS.LEVEL_02;

import java.util.Arrays;

public class Q49993 {

    public static void main(String[] args) {
        /*
        skill	skill_trees	                        return
        "CBD"	["BACDE", "CBADF", "AECB", "BDA"]	2
         */
        String skill = "CBD";
        String[] skill_trees = {"BACDE", "CBADF", "AECB", "BDA"};
        System.out.println(solution(skill, skill_trees));
    }

    public static int solution(String skill, String[] skill_trees) {

        int[] priority = new int['Z' - 'A' + 1];
        Arrays.fill(priority, -1);

        for (int i = skill.length() - 1; i > 0; --i) {
            priority[skill.charAt(i) - 'A'] = skill.charAt(i - 1) - 'A';
        }

        int answer = 0;
        for (String skillTree : skill_trees) {

            if (rightOrder(skillTree, priority)) {
                ++answer;
            }
        }
        return answer;
    }

    public static boolean rightOrder(String skillTree, int[] priority) {

        boolean[] check = new boolean[priority.length];
        for (int i = 0; i < skillTree.length(); ++i) {

            // 선행 스킬이 있는데 check 안되어있으면 false
            int idx = skillTree.charAt(i) - 'A';
            if (priority[idx] >= 0 && !check[priority[idx]]) {
                return false;
            }
            check[idx] = true;
        }
        return true;
    }
}
